slatedb/src/seq_tracker.rs

```diff
@@ -22,204 +29,243 @@ pub(crate) enum FindOption {
+        let mut new_timestamps = Vec::with_capacity(self.capacity as usize);
 
-        if seq % self.step != 0 {
-            return;
+        for i in (0..self.sequence_numbers.len()).step_by(2) {
+            new_seqs.push(self.sequence_numbers[i]);
+            new_timestamps.push(self.timestamps[i]);
         }
 
-        if self.timestamps.is_empty() {
-            self.first_seq = seq;
-        }
+        self.sequence_numbers = new_seqs;
+        self.timestamps = new_timestamps;
+    }
 
-        self.last_seq = seq;
-        self.timestamps.push(ts);
-
-        // if we've exceeded the capacity then downsample the
-        // sequence numbers and double the step size
-        if self.timestamps.len() as u32 > self.capacity {
-            let old = std::mem::take(&mut self.timestamps);
-            let mut downsampled = Vec::with_capacity(self.capacity as usize);
-            for (i, &ts) in old.iter().enumerate() {
-                if i % 2 == 0 {
-                    downsampled.push(ts);
-                }
-            }
-            self.timestamps = downsampled;
-            self.step *= 2;
+    /// Find the timestamp for a given sequence number
+    pub(crate) fn find_ts(&self, seq: u64, find_opt: FindOption) -> Option<DateTime<Utc>> {
+        if self.sequence_numbers.is_empty() {
+            return None;
         }
-    }
 
-    /// finds the timestamp associated with `seq`, rounding up or down
-    /// as specified by the find option if the sequence number does not
-    /// exist in this tier
-    fn find_ts(&self, seq: u64, find_opt: FindOption) -> Option<i64> {
-        let len = (self.timestamps.len()) as u64;
-        let max = self.first_seq + self.step * len;
-
-        if let Some(offset) = match find_opt {
-            FindOption::RoundUp if seq <= self.last_seq => {
-                if seq >= self.first_seq {
-                    Some(cmp::max(0, (seq - self.first_seq) as i64))
-                } else {
-                    Some(0)
+        match self.sequence_numbers.binary_search(&seq) {
+            Ok(idx) => DateTime::from_timestamp(self.timestamps[idx], 0),
+            Err(idx) => match find_opt {
+                FindOption::RoundUp if idx < self.sequence_numbers.len() => {
+                    DateTime::from_timestamp(self.timestamps[idx], 0)
                 }
-            }
-            FindOption::RoundDown if seq >= self.first_seq => {
-                Some(cmp::min((seq - self.first_seq) as i64, max as i64))
-            }
-            _ => None,
-        } {
-            let mut idx = offset.div_euclid(self.step as i64);
-            let remainder = offset.rem_euclid(self.step as i64);
-
-            if remainder != 0 && find_opt == FindOption::RoundUp {
-                idx += 1;
-            }
-
-            idx = idx.clamp(0, len as i64 - 1);
-            return Some(self.timestamps[idx as usize]);
+                FindOption::RoundDown if idx > 0 => {
```

Comment:
Trying to wrap my head around the edge cases here. If len == 1 and `FindOption::RoundDown`, we'll always return None? I guess that's right since you'd expect a lower value returned.

Reply:
don't forget the `Ok(idx)` branch! if there's an exact match, then it'll return the value, if there isn't and you asked it to round down then there's no value stored so we could either return `None` or some sentinel value like `0`. I figured `None` makes more sense.

Reply:
Yep, agreed. Just making sure I got it. :)